pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/operands/FunctionOperand.java

```diff
@@ -52,6 +52,9 @@ public FunctionOperand(RexExpression.FunctionCall functionCall, DataSchema dataS
     _resultName = computeColumnName(functionCall.getFunctionName(), _childOperandList);
     // TODO: Check type match between functionCall's data type and result type.
     _resultType = FunctionUtils.getColumnDataType(_functionInvoker.getResultClass());
+
+    boolean isTypeMatched = functionCall.getDataType() == _resultType.toDataType();
+    Preconditions.checkState(isTypeMatched, "Mismatch function data type and result type");
```

Comment:
Pinot's FunctionInvoker largely depending on the next operator's FunctionInvoker parameter type casting 
(see: https://github.com/apache/pinot/blob/d1227e466dc6e350214485334cdf4eedc7502d67/pinot-common/src/main/java/org/apache/pinot/common/function/FunctionInvoker.java#L115)
however we are not guaranteed to have another function invoker chained after.

For this i think we can use `functionCall.getDataType()` and try to use it to convert `_resultType` similar to the link above. (it would have performance issues but a good starting point)

please comment on the issue and ask for some specific repo examples so we can add to the test cases

Reply:
Thanks for the help. I have a couple of questions. 
- For conversion(`parameterType.convert`), do I have to convert the `_resultType`(`fieldSpec.DataType`)  to PinotDataType and perform the conversion or use the conversion function`FieldSpec.convertInternal` directly.
- Second, what should be the outcome of the failing `caswhen` query? It should fail, or with conversion, it should execute without error.

Thanks.

Reply:
from my perspective
- either should work

> Thanks for the help. I have a couple of questions.
> 
> * For conversion(`parameterType.convert`), do I have to convert the `_resultType`(`fieldSpec.DataType`)  to PinotDataType and perform the conversion or use the conversion function`FieldSpec.convertInternal` directly.

Either should work. but would go with the safest, which is what FunctionInvoker handles parameter type

> * Second, what should be the outcome of the failing `caswhen` query? It should fail, or with conversion, it should execute without error.

it should not fail at all at this stage --> calcite should ensure the resulting data type is consistent. otherwise it will fail in planner before reaching runtime. If the data is not what it said it is (from input) then the failure handling is the same as all other operators

Reply:
@walterddr Can you please take a look? Thanks

Reply:
thank you will do. i also asked for more examples from @ankitsultana in the original issue. let's wait for a bit and add more test cases

Reply:
I think it might be worthwhile to add a couple or so queries which do comparisons of double/float columns with double/float literals, and likewise for int/long columns.

Stuff like:

```
case longCol > 10
...
case doubleCol < 1.3
case floatCol < 10
..
where longCol < 10 and doubleCol < 1.3 and floatCol < 10...

..
with tmp as (select sum(floatCol) as float_sum from ... group by col1) select * from tmp where float_sum < 10
```

Reply:
I aI added a couple of test cases as suggested. Please let me know if I have to add more.

Reply:
looking